File: packages/aid/src/pka.vectors.test.ts
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { discover } from './index.js';
import * as fs from 'node:fs';
import * as path from 'node:path';
import * as nodeCrypto from 'node:crypto';

// Force DNS miss to drive well-known fallback path
vi.mock('dns-query', () => {
  return {
    query: vi.fn(async () => {
      const err: any = new Error('ENOTFOUND');
      err.code = 'ENOTFOUND';
      throw err;
    }),
  };
});

function loadVectors() {
  const p = path.resolve(process.cwd(), '..', '..', 'protocol', 'pka_vectors.json');
  const raw = fs.readFileSync(p, 'utf8');
  return JSON.parse(raw) as {
    version: number;
    vectors: Array<{
      id: string;
      desc: string;
      record: { v: string; u: string; p: string; i: string };
      key: { public: string; seed_b64: string };
      covered: string[];
      created: number;
      httpDate?: string;
      overrideAlg?: string;
      overrideKeyId?: string;
      expect: 'pass' | 'fail';
    }>;
  };
}

function seedToPkcs8Ed25519(seed: Uint8Array): Uint8Array {
  // PKCS#8 structure for Ed25519 private key (RFC 8410)
  // 30 2e 02 01 00 30 05 06 03 2b 65 70 04 22 04 20 || 32-byte seed
  const header = Uint8Array.from([
    0x30, 0x2e, 0x02, 0x01, 0x00, 0x30, 0x05, 0x06, 0x03, 0x2b, 0x65, 0x70, 0x04, 0x22, 0x04, 0x20,
  ]);
  const out = new Uint8Array(header.length + seed.length);
  out.set(header, 0);
  out.set(seed, header.length);
  return out;
}

function b58encode(bytes: Uint8Array): string {
  const ALPHABET = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';
  let zeros = 0;
  while (zeros < bytes.length && bytes[zeros] === 0) zeros++;
  const size = Math.ceil((bytes.length * Math.log(256)) / Math.log(58)) + 1;
  const b = new Uint8Array(size);
  let length = 0;
  for (let i = zeros; i < bytes.length; i++) {
    let carry = bytes[i];
    let j = size - 1;
    while (carry !== 0 || j >= size - length) {
      carry += 256 * b[j];
      b[j] = carry % 58;
      carry = Math.floor(carry / 58);
      j--;
    }
    length = size - 1 - j;
  }
  let it = size - length;
  while (it < size && b[it] === 0) it++;
  let out = '1'.repeat(zeros);
  for (let i = it; i < size; i++) out += ALPHABET[b[i]];
  return out;
}

function extractEd25519PublicFromSpki(spkiDer: Uint8Array): Uint8Array {
  // Ed25519 SPKI ends with BIT STRING of 0x21 0x00 + 32 bytes public key; extract last 32 bytes
  return spkiDer.subarray(spkiDer.length - 32);
}

describe('PKA vectors parity (TS)', () => {
  const g = globalThis as any;
  let origFetch: any;

  beforeEach(() => {
    origFetch = g.fetch;
  });
  afterEach(() => {
    g.fetch = origFetch;
    vi.restoreAllMocks();
  });

  for (const v of loadVectors().vectors) {
    it(`${v.id}: ${v.desc}`, async () => {
      const seed = Buffer.from(v.key.seed_b64, 'base64');
      const pkcs8 = seedToPkcs8Ed25519(seed);
      const priv = nodeCrypto.createPrivateKey({ key: pkcs8, format: 'der', type: 'pkcs8' });
      const pub = nodeCrypto
        .createPublicKey(priv)
        .export({ type: 'spki', format: 'der' }) as Buffer;
      const rawPub = extractEd25519PublicFromSpki(new Uint8Array(pub));
      const pka = 'z' + b58encode(rawPub);

      g.fetch = vi.fn(async (url: string, init?: { headers?: Record<string, string> }) => {
        if (url.includes('/.well-known/agent')) {
          return {
            ok: true,
            status: 200,
            headers: {
              get: (n: string) => (n.toLowerCase() === 'content-type' ? 'application/json' : null),
            },
            text: async () =>
              JSON.stringify({
                v: v.record.v,
                u: v.record.u,
                p: v.record.p,
                k: pka,
                i: v.record.i,
              }),
          };
        }
        // Handshake
        const order = v.covered;
        const challenge = init?.headers?.['AID-Challenge'] ?? '';
        const requestDate = init?.headers?.['Date'] ?? '';
        const date = v.expect === 'pass' ? requestDate : (v.httpDate ?? requestDate);
        const method = 'GET';
        const target = url;
        const host = new URL(url).host;
        const lines: string[] = [];
        for (const item of order) {
          switch (item) {
            case 'AID-Challenge':
              lines.push(`"AID-Challenge": ${challenge}`);
              break;
            case '@method':
              lines.push(`"@method": ${method}`);
              break;
            case '@target-uri':
              lines.push(`"@target-uri": ${target}`);
              break;
            case 'host':
              lines.push(`"host": ${host}`);
              break;
            case 'date':
              lines.push(`"date": ${date}`);
              break;
            default:
              throw new Error('unexpected covered item: ' + item);
          }
        }
        const keyid = v.overrideKeyId ?? v.record.i;
        const alg = v.overrideAlg ?? 'ed25519';
        const nowSec = Math.floor(Date.now() / 1000);
        const usedCreated = v.expect === 'pass' ? nowSec : v.created;
        const paramsStr = `(${order.map((c) => `"${c}"`).join(' ')});created=${usedCreated};keyid=${keyid};alg="${alg}"`;
        lines.push(`"@signature-params": ${paramsStr}`);
        const base = new TextEncoder().encode(lines.join('\n'));
        const sig = nodeCrypto.sign(null, base, priv);
        const headers = {
          get: (name: string) => {
            const k = name.toLowerCase();
            if (k === 'signature-input')
              return `sig=("${order.join('" "')}");created=${usedCreated};keyid=${keyid};alg="${alg}"`;
            if (k === 'signature') return `sig=:${Buffer.from(sig).toString('base64')}:`;
            if (k === 'date') return date;
            return null;
          },
        };
        return { ok: true, status: 200, headers, text: async () => '' };
      });

      if (v.expect === 'pass') {
        const { record } = await discover('example.com', { wellKnownFallback: true });
        expect(record.pka).toBeDefined();
      } else {
        await expect(discover('example.com', { wellKnownFallback: true })).rejects.toMatchObject({
          errorCode: 'ERR_SECURITY',
        });
      }
    });
  }
});

File: packages/aid-go/pka_vectors_test.go
package aid

import (
    "context"
    "crypto/ed25519"
    "encoding/base64"
    "encoding/json"
    "io"
    "net"
    "net/http"
    "os"
    "path/filepath"
    "strconv"
    "strings"
    "testing"
    "time"
)

type vector struct {
    ID      string `json:"id"`
    Desc    string `json:"desc"`
    Record  struct{
        V string `json:"v"`
        U string `json:"u"`
        P string `json:"p"`
        I string `json:"i"`
    } `json:"record"`
    Key     struct{ Public string `json:"public"`; SeedB64 string `json:"seed_b64"` } `json:"key"`
    Covered []string `json:"covered"`
    Created int64 `json:"created"`
    HTTPDate string `json:"httpDate"`
    OverrideAlg string `json:"overrideAlg"`
    OverrideKeyId string `json:"overrideKeyId"`
    Expect string `json:"expect"`
}

type vectorsFile struct{
    Version int `json:"version"`
    Vectors []vector `json:"vectors"`
}

// simple base58 encode for tests
func b58encode(data []byte) string {
    const alphabet = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"
    // count leading zeros
    zeros := 0
    for zeros < len(data) && data[zeros] == 0 { zeros++ }
    // approximate size
    size := len(data)*138/100 + 1
    b := make([]byte, size)
    for _, v := range data {
        carry := int(v)
        for j := size-1; j >=0; j-- {
            carry += 256 * int(b[j])
            b[j] = byte(carry % 58)
            carry /= 58
        }
    }
    // skip leading zeros in b
    it := 0
    for it < size && b[it] == 0 { it++ }
    out := strings.Repeat("1", zeros)
    for ; it < size; it++ { out += string(alphabet[b[it]]) }
    return out
}

// mockRoundTripper responds to well-known and handshake requests based on a vector
type mockRoundTripper struct{ V vector; Pka string }

func (m *mockRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
    if strings.HasSuffix(req.URL.Path, "/.well-known/agent") {
        // Return JSON with record and computed PKA
        body := []byte(`{"v":"` + m.V.Record.V + `","u":"` + m.V.Record.U + `","p":"` + m.V.Record.P + `","k":"` + m.Pka + `","i":"` + m.V.Record.I + `"}`)
        return &http.Response{StatusCode: 200, Header: http.Header{"Content-Type": []string{"application/json"}}, Body: io.NopCloser(strings.NewReader(string(body)))}, nil
    }
    // Handshake response
    challenge := req.Header.Get("AID-Challenge")
    reqDate := req.Header.Get("Date")
    // For passing cases, make created within window using current time, otherwise use vector's Created/HTTPDate
    created := m.V.Created
    date := reqDate
    if m.V.Expect == "pass" {
        created = time.Now().Unix()
        date = reqDate
    } else if m.V.HTTPDate != "" {
        date = m.V.HTTPDate
    }
    order := m.V.Covered
    keyid := m.V.OverrideKeyId
    if keyid == "" { keyid = m.V.Record.I }
    alg := m.V.OverrideAlg
    if alg == "" { alg = "ed25519" }

    // Build base per spec formatting
    var lines []string
    for _, item := range order {
        switch strings.ToLower(item) {
        case "aid-challenge": lines = append(lines, "\"AID-Challenge\": "+challenge)
        case "@method": lines = append(lines, "\"@method\": GET")
        case "@target-uri": lines = append(lines, "\"@target-uri\": "+req.URL.String())
        case "host": lines = append(lines, "\"host\": "+req.URL.Host)
        case "date": lines = append(lines, "\"date\": "+date)
        }
    }
    // Params (use keyid raw as provided)
    quoted := make([]string, 0, len(order))
    for _, c := range order { quoted = append(quoted, "\""+c+"\"") }
    params := "(" + strings.Join(quoted, " ") + ");created=" +  strconvI(created) + ";keyid=" + keyid + ";alg=\"" + alg + "\""
    lines = append(lines, "\"@signature-params\": "+params)
    base := []byte(strings.Join(lines, "\n"))

    // Sign base
    seed, _ := base64.StdEncoding.DecodeString(m.V.Key.SeedB64)
    sk := ed25519.NewKeyFromSeed(seed)
    sig := ed25519.Sign(sk, base)
    sigB64 := base64.StdEncoding.EncodeToString(sig)

    h := make(http.Header)
    h.Set("Signature-Input", "sig=("+strings.Join(quoted, " ")+");created="+strconvI(created)+";keyid="+keyid+";alg=\""+alg+"\"")
    h.Set("Signature", "sig=:"+sigB64+":")
    h.Set("Date", date)
    return &http.Response{StatusCode: 200, Header: h, Body: io.NopCloser(strings.NewReader(""))}, nil
}

func strconvI(i int64) string { return strconv.FormatInt(i, 10) }

func TestPKAVectors(t *testing.T) {
    // Load vectors
    // Relative path from this package to protocol/pka_vectors.json
    path := filepath.Join("..", "..", "protocol", "pka_vectors.json")
    raw, err := os.ReadFile(path)
    if err != nil { t.Skip("vectors not available: ", err) }
    var vf vectorsFile
    if err := json.Unmarshal(raw, &vf); err != nil { t.Fatalf("parse vectors: %v", err) }

    // Override DNS to force well-known
    oldLookup := lookupTXT
    defer func(){ lookupTXT = oldLookup }()
    lookupTXT = func(ctx context.Context, name string) ([]string, error) { return nil, &net.DNSError{Err:"no such host", Name: name} }

    for _, v := range vf.Vectors {
        // Compute PKA from seed
        seed, _ := base64.StdEncoding.DecodeString(v.Key.SeedB64)
        sk := ed25519.NewKeyFromSeed(seed)
        pub := sk.Public().(ed25519.PublicKey)
        pka := "z" + b58encode(pub)

        // Install mock HTTP client
        oldClient := httpClient
        httpClient = &http.Client{ Transport: &mockRoundTripper{ V: v, Pka: pka } }
        // Execute discovery (domain unused as we mock DNS and HTTP)
        _, _, err := Discover("example.com", 2*time.Second)
        // Restore client for next iteration
        httpClient = oldClient

        if v.Expect == "pass" && err != nil {
            t.Fatalf("%s expected pass, got error: %v", v.ID, err)
        }
        if v.Expect == "fail" && err == nil {
            t.Fatalf("%s expected fail, got success", v.ID)
        }
    }
}
